File: intervals/452. Minimum Number of Arrows to Burst Balloons/js/main.js
// Time O(NLogN)
// Space O(N)
const findMinArrowShots = points => {
  if (points.length == 0) {
    return 0;
  }

  points.sort((a, b) => a[0] - b[0]);

  let [start, end] = points[0];
  let res = 1;

  for (let i = 1; i < points.length; i++) {
    if (overlap([start, end], points[i])) {
      start = Math.max(start, points[i][0]);
      end = Math.min(end, points[i][1]);
    } else {
      res++;
      start = points[i][0];
      end = points[i][1];
    }
  }

  return res;

  function overlap([x, y], [u, z]) {
    return z >= x && y >= u;
  }
};

File: intervals/452. Minimum Number of Arrows to Burst Balloons/py/main.py
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        if not points:
            return 0
        
        points.sort(key = lambda x: x[0])
        
        res = 1
        start, end = points[0]
                
        for i in range(1, len(points)):
            if self.overlap([start, end], points[i]):
                start = max(start, points[i][0])
                end = min(end, points[i][1])
            else:
                res += 1
                start, end = points[i]
        return res
            
    def overlap(self, x, y):
        return y[1] >= x[0] and x[1] >= y[0]